[WIP] Allow dotted access to yaml-loaded dicts: d.key1.key2.key3

rhaschke opened the pull request:
Implements #240. Requires some unit test...

rhaschke committed:
Allow dotted access to yaml-loaded dicts: d.key1.key2.key3

gavanderhoorn commented:
This implementation basically works, with one 'problem': the following syntax doesn't work:

```xml
<xacro:property name="foo" value="${my_yaml.blop.flip['some_key'].baz}" />
```

Reason this doesn't work is that `flip['some_key']` returns a regular `dict`, not a `YamlDictWrapper`, so the `.` operator doesn't work.

If we feel that is a problem, I can change it, but I wanted to see whether we do (feel that is a problem).

---

Edit: and another issue of course is that anything that's not a valid Python identifier (so something with a space in it fi) or a reserved word will cause problems.

That's a limitation I believe is fair though and we can document.

gavanderhoorn commented on src/xacro/__init__.py:
```diff
@@ -94,6 +94,18 @@ def abs_filename_spec(filename_spec):
+        result = self.__d.__getitem__(item)
+        return YamlDictWrapper(result) if isinstance(result, dict) else result
+
+    def __getitem__(self, item):
+        return self.__d.__getitem__(item)
+
```
Did you have a particular reason to not just `__getitem__ = __getattr__`?

rhaschke replied:
No, I just didn't think about that. This would solve the "problem" you mentioned of course. I'm fine with this change.

gavanderhoorn replied:
I'll check how that works with non-existing keys. I remember there being some difference in the behaviour between `__getitem__` and `__getattr__`.

gavanderhoorn committed:
dotted-dict: alias getitem to getattr.

We need both to behave the same (ie: return a YamlDictWrapper when appropriate).

gavanderhoorn committed:
dotted-dict: raise slightly more informative exception on KeyError.

gavanderhoorn committed:
dotted-dict: add some tests.

gavanderhoorn commented:
@rhaschke: I've submitted ubi-agni/xacro#3 against your fork.

Let me know whether you deem this sufficient.

gavanderhoorn commented:
Oh, and would you consider this for a backport to Melodic?

gavanderhoorn commented:
And something else to consider: would we want to make the `YamlDictWrapper` read-only?

I'm not entirely sure whether we could assign something in an expression in `xacro`, but it would be prudent perhaps to explicitly make it impossible in `YamlDictWrapper`.

rhaschke commented:
Thanks a lot. I will have a look at your contributions over the weekend.

> Would we want to make the YamlDictWrapper read-only?

Isn't it read-only already, because we don't provide `__setitem__` or `__setattr__`?

> Would you consider this for a backport to Melodic?

Yes, in principle no problem. I was fearing some side effects or name conflicts initially. But, as the property return by load_yaml is unique anyway, there should be a problem.

gavanderhoorn commented:
> > Would we want to make the YamlDictWrapper read-only?
> 
> Isn't it read-only already, because we don't provide `__setitem__` or `__setattr__`?

True.

> > Would you consider this for a backport to Melodic?
> 
> Yes, in principle no problem. I was fearing some side effects or name conflicts initially. But, as the property return by load_yaml is unique anyway, there should be a problem.

That should be "there shouldn't be a problem", correct?

rhaschke commented:
Yes, sorry. There should**n't** be a problem. :rofl:

gavanderhoorn commented:
Nice.

Thanks @rhaschke.

rhaschke commented:
I rebased this onto melodic-devel. Prior to a new release, I want to improve warning messages (#236).

gavanderhoorn commented:
Ok.

Throwing the `XacroException` on the `KeyError` in the `YamlDictWrapper` makes for quite OK error message though ("no such key 'X', while evaluation expression Y").

But I guess this is more about the general way errors are reported?

gavanderhoorn commented:
Which branch did this get merged into?

The PR states `noetic-devel`, but I can't find the commits there?

Did it end up in `ubi-agni/xacro` (https://github.com/ubi-agni/xacro/commit/d1171d06d0ebef15eac246ca1b6ef24c3314c9c1)?

rhaschke commented:
Sorry. Indeed, I missed to push to both, ubi-agni/xacro and ros/xacro. Fixed now.

gavanderhoorn commented:
Thanks.

Noetic will get it later?

rhaschke commented:
Yes. I will merge forward changes from Melodic to Noetic after the release. This results in a much cleaner history than, e.g. in MoveIt.

alikureishy commented on src/xacro/__init__.py:
```diff
@@ -104,7 +116,7 @@ def load_yaml(filename):
     oldstack = push_file(filename)
     try:
-        return yaml.safe_load(f)
+        return YamlDictWrapper(yaml.safe_load(f))
```
Could YamlDictWrapper be exposed for general use in Xacro files? Why restrict the dot notation to just Yaml-loaded files? Why not allow any dictionary to be wrapped with this to enable read-only dotted access? For example:

```
<xacro:property name="dictionary" value="${dict(a=1, b=2}"/>
<xacro:property name="dotted" value="${YamlDictWrapper(dictionary)}"/>
```

However, if I try this directly, I get an error:
` name 'YamlDictWrapper' is not defined`

To make it even more user-friendly, perhaps a function 'dotify' could be exposed as below, that wraps the given dictionary within YamlDictWrapper?
`<xacro:property name="dotted" value="${dotify(dictionary)}"/>
`

rhaschke replied:
That should be possible. Other opinions, @gavanderhoorn, particularly about the function name `dotify`?

gavanderhoorn replied:
Hm.

I'm not sure about the name `dotify`. Seems to be meaningful only if you already know what "dotted access" is.

Can't really come up with anything better right now though.

Also: to avoid breaking existing user's setups: would it make sense to add some sort of prefix? I'm not sure whether that's actually a risk, but seeing as `xacro` is such a common dependency, it would be good to avoid the possibility.

rhaschke replied:
I'm not good at inventing new names either. What kind of prefix do you think of? Something like `xacro.dotify`?
And we warn/throw if the user attempts to override the namespace `xacro`?

gavanderhoorn replied:
> Something like `xacro.dotify`?

yes, that might be nice actually.

> And we warn/throw if the user attempts to override the namespace `xacro`?

that would seem prudent, yes.

rhaschke replied:
I'm hesitant to introduce a `xacro` package namespace. Actually, we already have a safety check issuing a warning if any of our global symbols is overridden: 
https://github.com/ros/xacro/blob/a4f0b84e166de5742f6ac83c561d1e5b92eb3ba7/src/xacro/__init__.py#L325-L326

As overriding works, it will not break existing xacros as well (just issue the warning).
